source/common/common/utility.cc

```diff
@@ -33,6 +117,41 @@ std::string DateFormatter::fromTime(time_t time) const {
   return std::string(&buf[0]);
 }
 
+std::string
+DateFormatter::fromTimeAndPrepareSubsecondOffsets(time_t time,
+                                                  SubsecondOffsets& subsecond_offsets) const {
+  tm current_tm;
+  gmtime_r(&time, &current_tm);
+
+  std::array<char, 1024> buf;
+  std::string formatted;
+
+  size_t step = 0;
+  int32_t previous = 0;
+  subsecond_offsets.reserve(subseconds_.size());
+  for (const auto& subsecond : subseconds_) {
+    ASSERT(step < format_string_.size());
+    const std::string segment = format_string_.substr(step, subsecond.position_ - step);
+    const size_t formatted_length = strftime(&buf[0], buf.size(), segment.c_str(), &current_tm);
+    absl::StrAppend(&formatted, &buf[0],
+                    format_string_.substr(subsecond.position_, subsecond.width_));
+
+    const int32_t offset = formatted_length - segment.size();
+    subsecond_offsets.emplace_back(previous + offset);
+
+    step = subsecond.position_ + subsecond.width_;
+    previous += offset;
+  }
+
+  if (step < format_string_.size()) {
+    const std::string last_segment = format_string_.substr(step);
+    strftime(&buf[0], buf.size(), last_segment.c_str(), &current_tm);
+    absl::StrAppend(&formatted, &buf[0]);
+  }
```

Comment:
Wow, yeah, this is complicated, but I'm not sure if there's a simpler way. I'm OK with the above if you clean it up and add comments. Alternatively, we could fragment into segments during the initial parse (which I think is what you may have had in an earlier iteration) and just do the append + substitute here.

Reply:
I do the pre-computation 96f915e.